test/extensions/filters/http/cdn_loop/filter_integration_test.cc

```diff
@@ -0,0 +1,187 @@
+  ASSERT_TRUE(response->complete());
+  EXPECT_EQ("400", response->headers().getStatusValue());
+}
+
+TEST_P(CdnLoopFilterIntegrationTest, CdnLoop2Allowed1Seen) {
+  config_helper_.addFilter(MaxOf2Config);
+  initialize();
+  codec_client_ = makeHttpConnection(makeClientConnection(lookupPort("http")));
+
+  Http::TestRequestHeaderMapImpl request_headers{{":method", "GET"},
+                                                 {":path", "/"},
+                                                 {":scheme", "http"},
+                                                 {":authority", "host"},
+                                                 {"CDN-Loop", "cdn"}};
+
+  auto response = sendRequestAndWaitForResponse(request_headers, 0, default_response_headers_, 0);
+
+  const auto* payload_entry = upstream_request_->headers().get(Http::LowerCaseString("CDN-Loop"));
+  ASSERT_NE(payload_entry, nullptr);
+  EXPECT_EQ(payload_entry->value().getStringView(), "cdn,cdn");
+  ASSERT_TRUE(response->complete());
+  EXPECT_EQ("200", response->headers().getStatusValue());
+}
+
+TEST_P(CdnLoopFilterIntegrationTest, CdnLoop2Allowed2Seen) {
+  config_helper_.addFilter(MaxOf2Config);
+  initialize();
+  codec_client_ = makeHttpConnection(makeClientConnection(lookupPort("http")));
+
+  Http::TestRequestHeaderMapImpl request_headers{{":method", "GET"},
+                                                 {":path", "/"},
+                                                 {":scheme", "http"},
+                                                 {":authority", "host"},
+                                                 {"CDN-Loop", "cdn, cdn"}};
+
+  auto response = sendRequestAndWaitForResponse(request_headers, 0, default_response_headers_, 0);
+
+  const auto* payload_entry = upstream_request_->headers().get(Http::LowerCaseString("CDN-Loop"));
+  ASSERT_NE(payload_entry, nullptr);
+  EXPECT_EQ(payload_entry->value().getStringView(), "cdn, cdn,cdn");
+  ASSERT_TRUE(response->complete());
+  EXPECT_EQ("200", response->headers().getStatusValue());
+}
+
+TEST_P(CdnLoopFilterIntegrationTest, CdnLoop2Allowed3Seen) {
+  config_helper_.addFilter(MaxOf2Config);
+  initialize();
+  codec_client_ = makeHttpConnection(makeClientConnection(lookupPort("http")));
+
+  Http::TestRequestHeaderMapImpl request_headers{{":method", "GET"},
+                                                 {":path", "/"},
+                                                 {":scheme", "http"},
+                                                 {":authority", "host"},
+                                                 {"CDN-Loop", "cdn, cdn, cdn"}};
+
+  auto response = codec_client_->makeHeaderOnlyRequest(request_headers);
+  response->waitForEndStream();
+  ASSERT_TRUE(response->complete());
+  EXPECT_EQ("502", response->headers().getStatusValue());
+}
+
+INSTANTIATE_TEST_SUITE_P(Protocols, CdnLoopFilterIntegrationTest,
+                         testing::ValuesIn(HttpProtocolIntegrationTest::getProtocolTestParams()),
```

Comment:
I think we really only need to test downstream protocols here, not the full mesh of downstream + upstream, WDYT?

testing::ValuesIn(HttpProtocolIntegrationTest::getProtocolTestParams(
                             {Http::CodecClient::Type::HTTP1, Http::CodecClient::Type::HTTP2},
                             {FakeHttpConnection::Type::HTTP1})),

Or maybe even adding a getProtocolTestParams which just gets downstream protocols.

Reply:
The integration test doesn't take that long to run (~15s), so I'm happy to keep all 8 combinations if only to not have to answer the question "should we be testing a fake upstream with HTTP1 or HTTP2?".

Reply:
Yeah, unfortunately it's less about speed and more about CI resources (which without the infinite scalability of blaze we hit regularly).  I think there's value in the downstream variety because it tests header coalescing, but for upstream it doesn't add value and it's just one more straw on the metaphorical CI-camel's back.

Reply:
I'm convinced by CI frugality argument.  I've made the change!